core/src/main/java/org/apache/calcite/rel/metadata/RelMdMinRowCount.java

```diff
@@ -54,14 +54,31 @@
   }
 
   public Double getMinRowCount(Union rel, RelMetadataQuery mq) {
-    double rowCount = 0.0;
-    for (RelNode input : rel.getInputs()) {
-      Double partialRowCount = mq.getMinRowCount(input);
-      if (partialRowCount != null) {
-        rowCount += partialRowCount;
+    if (rel.all) {
+      double rowCount = 0.0;
+      for (RelNode input : rel.getInputs()) {
+        Double partialRowCount = mq.getMinRowCount(input);
+        if (partialRowCount != null) {
+          rowCount += partialRowCount;
+        }
+      }
+      return rowCount;
+    } else {
+      boolean valid = false;
+      double rowCount = Double.POSITIVE_INFINITY;
+      for (RelNode input : rel.getInputs()) {
+        Double partialRowCount = mq.getMinRowCount(input);
+        if (partialRowCount != null && rowCount > partialRowCount) {
+          rowCount = partialRowCount;
```

Comment:
@chunweilei  If I understand correctly this is basically trying to find the input with least number of rows, and that number is being used as minimum row count. If so then this looks wrong, because minimum number of rows for UNION DISTINCT will be 1 since the input with least number of rows can have all duplicates in which case only one row will be returned.

Reply:
Good point! PR was updated.